cprogs/xduplic-copier.1
.TH WITH-LOCK-EX "1" "July 2003" "Debian" "Chiark-utils-bin"
.SH NAME
xduplic-copier \- type into multiple X windows at once
.SH SYNOPSIS
.B xduplic-copier
.SH DESCRIPTION
xduplic-copier lets you type into multiple X windows at once.

It has a very basic user interface for selecting which windows to type
into.
.SH OPTIONS
xduplic-copier ignores its command-line arguments.
.SH MODES
xduplic-copier puts up a small window with some text in it.  It has
two modes:
.TP
Idle: "\fBi 0\fR" (red); Typing "\fBT\fR \fIcount\fR" (green)
If you type keystrokes into xduplic-copier, they will be replicated to
all the selected windows (if there are any).  (You can also type into
the selected windows individually in the normal way.)
count is the number of selected windows.
The starting mode is Idle; you should left-click to start selecting.
.TP
Selecting: "\fBS\fR \fIcount\fR" (white)
In this mode xduplic-copier has grabbed the mouse pointer and you
can indicate which windows you are going to want to type into.
.SH MOUSE ACTIONS
.TP
Left-click in xduplic-copier
Switch between typing and selecting modes.
.TP
Right-click in xduplic-copier
Quit.
.TP
Left-click in another window while selecting
Select this window.  (Beeps if already selected.)
.TP
Right-click in another window while selecting
Deselect this window.  (Beeps if not selected.)
.TP
Right-click in root window (ie, desktop background) while selecting
Deselect all windows.
.TP
Q key while selecting
Quits.
.SH XTERM AND ALLOW SENDEVENTS
xduplic-copier does its work by generating synthetic events for the
selected windows, using XSendEvent.  Unfortunately the xterm authors
think that allowing XSendEvent is a security problem; they are wrong:
any untrusted person can already take over your xterms anyway.  But
the xterm authors have configured xterm to discard synthetic events by
default.

You can solve this at runtime by bringing up the ctrl-leftbutton menu in
each xterm, and ticking the option "Allow SendEvents".  Or you can set
the allowSendEvents in your xterm X resources.

Neither of these significantly reduce your security.  Indeed, xterm
itself allows these properties to be set via the X toolkit system's
remote widget property setting arrangements - so it would be possible
for xduplic-copier to set this property itself on any xterms it
encountered.  Unfortunately doing so would be a lot of tedious
programming.
.SH BUGS
If one of the windows you have selected is closed, and you try typing,
xduplic-copier will crash due to an unhandled X11 error.  If you
notice that you have got into this state, you cannot retain your set
of selected windows because there is no way to click on the
now-destroyed window to deselect it.  You can keep xduplic-copier from
crashing by right-clicking on the background in selecting mode, and
then reselecting all your windows, which may be marginally more
convenient than restarting it.

The UI is perhaps excessively sparse.
.SH AUTHOR AND COPYRIGHT
Ian Jackson <ijackson@chiark.greenend.org.uk> wrote xduplic-copier
some time in 2002, and updated it in 2013. The manpage is from 2013.

xduplic-copier is govered by the GNU GPL, v3 or later.


debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/xduplic-copier.c
/*
 * Copyright (C) 2002,2013 Ian Jackson <ian@chiark.greenend.org.uk>
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <X11/Xlib.h>
#include <X11/keysym.h>
#include <X11/cursorfont.h>
#include <X11/cursorfont.h>
#include <X11/Xmu/WinUtil.h>
#include <X11/XKBlib.h>

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <unistd.h>

static Display *display;
static int selecting, l1_x, l1_y;
static char stringbuf[20];
static unsigned long c_black, c_white, c_red, c_yellow;
static Window w, root;
static Cursor cursor;
static GC gc;
static Colormap cmap;

struct wnode {
  struct wnode *next;
  Window w;
} *headwn;

static unsigned long getcolour(const char *name, int def) {
  Status st;
  XColor screen_def, exact_def;
  st= XAllocNamedColor(display,cmap,name,&screen_def,&exact_def);
  fprintf(stdout,"name %s pixel %lu\n",name,screen_def.pixel);
  return st ? screen_def.pixel : def;
}

static void beep(void) { XBell(display,100); }

/*
 *  While selecting:
 *                   Left                         Right
 *    application     select                       deselect
 *    root            start typing                 deselect all
 *    xduplic         start typing                 quit
 *
 *  While typing:
 *                   Left                         Right
 *    xduplic         start selecting              quit
 *
 *  Colours:
 *
 *    While typing:                   yellow on black
 *    While selecting:                white on black
 *    Idle (typing into nowhere):     red on black
 */

static void redisplay(void) {
  XClearWindow(display,w);
  XDrawString(display,w,gc, l1_x,l1_y, stringbuf,strlen(stringbuf));
}

static void restatus(void) {
  XGCValues v;
  int count;
  struct wnode *own;

  v.foreground= (selecting ? c_white :
		 headwn ? c_yellow :
		 c_red);
  
  XChangeGC(display,gc,
	    GCForeground,
	    &v);
  XClearWindow(display,w);

  for (count=0, own=headwn; own; own=own->next) count++;
  snprintf(stringbuf,sizeof(stringbuf),
	   "%c %d",
	   selecting ? 'S' :
	   headwn ? 'T' : 'i',
	   count);
  redisplay();
}

static void stopselecting(void) {
  XUngrabPointer(display,CurrentTime);
  selecting= 0;
  restatus();
}

static void startselecting(void) {
  Status st;
  st= XGrabPointer(display,root,True,
		   ButtonPressMask,GrabModeAsync,
		   GrabModeAsync,None,cursor,CurrentTime);
  if (st != Success) beep();
  else selecting= 1;
  restatus();
}

static void buttonpress(XButtonEvent *e) {
  struct wnode *own, **ownp, *ownn;
  int rightbutton;
  Window sw;

  switch (e->button) {
  case Button1: rightbutton=0; break;
  case Button3: rightbutton=1; break;
  default: return;
  }

  fprintf(stdout,"button right=%d in=%lx sub=%lx (w=%lx root=%lx)\n",
	  rightbutton, (unsigned long)e->window, (unsigned long)e->subwindow,
	  (unsigned long)w, (unsigned long)e->root);

  if (e->window == w) {
    if (rightbutton) _exit(0);
    if (selecting) {
      stopselecting();
      /* move pointer to where it already is, just in case wm is confused */
      XWarpPointer(display,None,root, 0,0,0,0, e->x_root,e->y_root);
    } else {
      startselecting();
    }
    return;
  }

  if (!selecting) return;

  if (e->window != e->root) return;

  if (!e->subwindow) {
    if (!rightbutton) {
      stopselecting();
    } else {
      if (!headwn) { beep(); return; }
      for (own=headwn; own; own=ownn) {
	ownn= own->next;
	free(own);
      }
      headwn= 0;
      restatus();
    }
    return;
  }

  sw= XmuClientWindow(display, e->subwindow);

  if (sw == w) { beep(); return; }

  for (ownp=&headwn;
       (own=(*ownp)) && own->w != sw;
       ownp= &(*ownp)->next);
  
  if (!rightbutton) {
    
    if (own) { beep(); return; }
    own= malloc(sizeof(*own)); if (!own) { perror("malloc"); exit(-1); }
    own->w= sw;
    own->next= headwn;
    headwn= own;

  } else {

    if (!own) { beep(); return; }
    *ownp= own->next;
    free(own);

  }

  restatus();
}

static void keypress(XKeyEvent *e) {
  Status st;
  struct wnode *own;
  unsigned long mask;
  
  if (selecting) {
    fprintf(stdout,"key type %d serial %lu (send %d) "
	    "window %lx root %lx sub %lx time %lx @%dx%d (%dx%dabs) "
	    "state %x keycode %u same %d\n",
	    e->type, e->serial, (int)e->send_event,
	    (unsigned long)e->window,
	    (unsigned long)e->root,
	    (unsigned long)e->subwindow,
	    (unsigned long)e->time,
	    e->x,e->y, e->x_root,e->y_root,
	    e->state, e->keycode, (int)e->same_screen);
    if (XkbKeycodeToKeysym(display, e->keycode, 0, 0) == XK_q) _exit(1);
    beep(); return;
  }
  for (own=headwn; own; own=own->next) {
    mask= (e->type == KeyPress ? KeyPressMask :
	   e->type == KeyRelease ? KeyReleaseMask :
	   KeyPressMask|KeyReleaseMask);
    e->window= own->w;
    e->subwindow= None;
    e->send_event= True;
    st= XSendEvent(display,own->w,True,mask,(XEvent*)e);
    if (st != Success) {
      fprintf(stdout,"sendevent to %lx %d mask %lx\n",
	      (unsigned long)own->w, st, mask);
    }
  }
}

static void expose(XExposeEvent *e) {
  if (e->count) return;
  redisplay();
}

int main(int argc, const char **argv) {
  XEvent e;
  XGCValues gcv;
  XSetWindowAttributes wv;
  int screen, direction, ascent, descent, l1_width, l1_height;
  XCharStruct overall;
  Font font;

  display= XOpenDisplay(0);
  if (!display) { fputs("XOpenDisplay failed\n",stderr); exit(-1); }
  screen= DefaultScreen(display);
  cmap= DefaultColormap(display,screen);
  root= DefaultRootWindow(display);
  
  c_black=   getcolour("black",  0);
  c_white=   getcolour("white",  1);
  c_yellow=  getcolour("yellow", c_white);
  c_red=     getcolour("red",    c_white);

  cursor= XCreateFontCursor(display,XC_crosshair);

  wv.event_mask= KeyPressMask|KeyReleaseMask|ButtonPressMask|ExposureMask;
  w= XCreateWindow(display, root,
		   0,0, 50,21, 0,DefaultDepth(display,screen),
		   InputOutput, DefaultVisual(display,screen),
		   CWEventMask, &wv);

  font= XLoadFont(display,"fixed");

  gcv.background= c_black;
  gcv.font=       font;
  gc= XCreateGC(display,w,GCBackground|GCFont,&gcv);

  XQueryTextExtents(display,font, "SIT 0689", 8,
		    &direction,&ascent,&descent,&overall);
  l1_width= overall.lbearing + overall.rbearing;
  l1_x= overall.lbearing;
  l1_y= ascent;
  l1_height= descent+ascent;

  XResizeWindow(display,w, l1_width,l1_height);
  XSetWindowBackground(display,w,c_black);
  
  XMapWindow(display,w);
  restatus();
  
  for (;;) {
    XNextEvent(display,&e);
    fprintf(stdout,"selecting = %d; event type = %lu\n",
	    selecting, (unsigned long)e.type);
    switch (e.type) {
    case Expose:                     expose(&e.xexpose);       break;
    case ButtonPress:                buttonpress(&e.xbutton);  break;
    case KeyPress: case KeyRelease:  keypress(&e.xkey);        break;
    }
  }
}
